libshaderc/src/shaderc_cpp_test.cc

```diff
@@ -504,6 +505,106 @@ TEST_F(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
               HasSubstr("void main(){ }"));
 }
 
+// A mock class that simulate a client of the includer.
+class TestIncluder : public shaderc::CompileOptions::Includer {
+ public:
+  // Create a fake file object, with specified path and content.
+  struct FakeFile {
+    std::string path;
+    std::string content;
+  };
+  // Use hashmap as a fake file system to store fake files to be included.
+  using FakeFS = std::unordered_map<std::string, FakeFile>;
+
+  TestIncluder(FakeFS& fake_fs) : fake_fs_(fake_fs){};
+
+  // Get path and content from the fake file system.
+  shaderc_includer_response* GetIncluderResponse(
+      const char* filename) override {
+    FakeFile& file_to_include = fake_fs_[filename];
+    response_.path = file_to_include.path.c_str();
+    response_.path_length = file_to_include.path.size();
+    response_.content = file_to_include.content.c_str();
+    response_.content_length = file_to_include.content.size();
+    return &response_;
+  };
+
+  // Response data is owned as private property, no need to release explicitly.
+  void ReleaseIncluderResponse(shaderc_includer_response* data) override {
+    (void)data;
+  };
+
+ private:
+  FakeFS& fake_fs_;
+  // Includer response data is stored as private property.
+  shaderc_includer_response response_;
+};
+
+TEST_F(CppInterface, Includer) {
+  const std::string kMainIncludingShader =
+      "void foo() {}\n"
+      "#include \"file_0\"\n"
+      "#include \"file_1\"\n";
+
+  TestIncluder::FakeFS fake_fs(
+      {{"file_0", {"path/to/file_0", "void bar() {}\n"}},
+       {"file_1", {"path/to/file_1", "void main() {foo(); bar();}\n"}}});
+
+  options_.SetIncluder(
+      std::unique_ptr<TestIncluder>(new TestIncluder(fake_fs)));
+
+  // Expects the compilation to succeed.
+  EXPECT_TRUE(CompilesToValidSpv(kMainIncludingShader,
+                                 shaderc_glsl_vertex_shader, options_));
+
+  // Sets compiler to preprocessing only mode, so we can check the result after
+  // preprocessing. Expect to see the content of the fake file object in the
+  // output of preprocessing.
+  options_.SetPreprocessingOnlyMode();
+  std::string preprocessing_output = CompilationOutput(
+      kMainIncludingShader, shaderc_glsl_vertex_shader, options_);
+  EXPECT_THAT(preprocessing_output, HasSubstr("#line 0 \"path/to/file_0\"\n"
+                                              " void bar(){ }\n"
```

Comment:
It seems to me all the cases can just reuse the same test code, parameterized by the fake filesystem contents, the name of an initial file (or its contents), and  the expected preprocessor output.
It would be easier to read the tests and see if they are covering the entire space we want.